fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java

```diff
@@ -4484,6 +4486,147 @@ public BigDecimal getProposedPrincipal() {
+
+        accountingBridgeDataAfterChargeOff.put("newLoanTransactions", newLoanTransactionsAfterChargeOff);
+        accountingBridgeData.add(accountingBridgeDataAfterChargeOff);
+
+        return accountingBridgeData;
+    }
+
+    private void getTransactionsForAccountingBridgeDataOnChargeOffDate(String currencyCode, List<Long> existingTransactionIds,
+            List<Long> existingReversedTransactionIds, List<Map<String, Object>> newLoanTransactionsBeforeChargeOff,
+            List<Map<String, Object>> newLoanTransactionsAfterChargeOff, Predicate<LoanTransaction> chargeOffDateCriteria) {
+        List<LoanTransaction> transactionsOnChargeOffDate = this.loanTransactions.stream().filter(chargeOffDateCriteria)
+                .collect(Collectors.toList());
+        /**
+         *
+         * TODO: Modify logic to retrieve correct charge-off transaction once reverse replay of charge-off is
+         * implemented
+         */
+        LoanTransaction chargeOffTransaction = this.loanTransactions.stream().filter(LoanTransaction::isChargeOff).findFirst().get();
+        Predicate<LoanTransactionRelation> isReplayed = transactionRelation -> transactionRelation.getRelationType()
+                .equals(LoanTransactionRelationTypeEnum.REPLAYED);
+        for (final LoanTransaction transaction : transactionsOnChargeOffDate) {
+            if (!transaction.isChargeOff()) {
+                if (transaction.isReversed() && existingTransactionIds.contains(transaction.getId())
+                        && !existingReversedTransactionIds.contains(transaction.getId())) {
+                    // transaction on charge-off date got reversed
+                    compareWithChargeOffIdAndAddTransactionForAccountingData(currencyCode, newLoanTransactionsBeforeChargeOff,
+                            newLoanTransactionsAfterChargeOff, transaction, chargeOffTransaction.getId());
+
+                } else if (!existingTransactionIds.contains(transaction.getId())) {
+                    // new transaction,backdated transaction or reverse replay on chargeoff date
+
+                    // check transaction relations
+                    if (!transaction.getLoanTransactionRelations().isEmpty()) {
+                        // if relations has REPLAYED relation then it is reverse replay scenario
+                        List<LoanTransactionRelation> replayedTransactionRelations = transaction.getLoanTransactionRelations().stream()
+                                .filter(isReplayed).collect(Collectors.toList());
+                        if (!replayedTransactionRelations.isEmpty()) {
+                            // get transaction Id for first reversed transaction relation
+                            Long transactionIdForReversedTransaction = replayedTransactionRelations.get(0).getToTransaction().getId();
+                            if (transactionIdForReversedTransaction > chargeOffTransaction.getId()) {
```

Comment:
I'm not sure about this, but are we using transaction id here, to determine whether the transaction happened before/after the chargeOff? I saw transactionDate in the code above, I think that is a better approach to use that. 
But I'm not familiar with the business logic here. Maybe @adamsaghy could help me with this.

Reply:
Both of them are used, the problem is only the transaction date precision is not enough (multiple transaction on the same day). By leveraging on the id is a workaround (for now)